src/wallet/scriptpubkeyman.cpp

```diff
@@ -2024,56 +2024,24 @@ isminetype DescriptorScriptPubKeyMan::IsMine(const CScript& script) const
 
 bool DescriptorScriptPubKeyMan::CheckDecryptionKey(const CKeyingMaterial& master_key, bool accept_no_keys)
 {
-    LOCK(cs_desc_man);
-    if (!m_map_keys.empty()) {
-        return false;
-    }
-
-    bool keyPass = m_map_crypted_keys.empty(); // Always pass when there are no encrypted keys
-    bool keyFail = false;
-    for (const auto& mi : m_map_crypted_keys) {
-        const CPubKey &pubkey = mi.second.first;
-        const std::vector<unsigned char> &crypted_secret = mi.second.second;
-        CKey key;
-        if (!DecryptKey(master_key, crypted_secret, pubkey, key)) {
-            keyFail = true;
-            break;
-        }
-        keyPass = true;
-        if (m_decryption_thoroughly_checked)
-            break;
-    }
-    if (keyPass && keyFail) {
-        LogPrintf("The wallet is probably corrupted: Some keys decrypt but not all.\n");
-        throw std::runtime_error("Error unlocking wallet: some keys decrypt but not all. Your wallet file may be corrupt.");
-    }
-    if (keyFail || (!keyPass && !accept_no_keys)) {
-        return false;
-    }
-    m_decryption_thoroughly_checked = true;
-    return true;
+    return m_keyman.CheckDecryptionKey(master_key, accept_no_keys);
 }
 
 bool DescriptorScriptPubKeyMan::Encrypt(const CKeyingMaterial& master_key, WalletBatch* batch)
 {
-    LOCK(cs_desc_man);
-    if (!m_map_crypted_keys.empty()) {
+    LOCK2(cs_desc_man, m_keyman.cs_keyman);
+
+    if (!m_keyman.Encrypt(master_key, batch)) {
         return false;
     }
 
-    for (const KeyMap::value_type& key_in : m_map_keys)
-    {
-        const CKey &key = key_in.second;
-        CPubKey pubkey = key.GetPubKey();
-        CKeyingMaterial secret(key.begin(), key.end());
-        std::vector<unsigned char> crypted_secret;
-        if (!EncryptSecret(master_key, secret, pubkey.GetHash(), crypted_secret)) {
-            return false;
-        }
-        m_map_crypted_keys[pubkey.GetID()] = make_pair(pubkey, crypted_secret);
-        batch->WriteCryptedDescriptorKey(GetID(), pubkey, crypted_secret);
+    for (const CKeyID& id : m_set_stored_keys) {
+        const auto& ckey_pair = m_keyman.GetCryptedKey(id);
+        assert(ckey_pair != std::nullopt);
```

Comment:
```suggestion
        const auto& ckey_pair = Assert(m_keyman.GetCryptedKey(id));
```

Reply:
Assertions shouldn't have side effects, hence the separation.